src/mds/MDSContext.cc

```diff
@@ -28,6 +28,9 @@ void MDSInternalContextBase::complete(int r) {
   assert(mds != NULL);
   assert(mds->mds_lock.is_locked_by_me());
   MDSContext::complete(r);
+
+  // avoid unhealthy heartbeat map when finish_contexts() processes large context list
+  mds->heartbeat_reset();
```

Comment:
This commit should be a separate PR.

Also, if we are processing a context list long enough to cause the heartbeat to timeout, then we are also not serving incoming requests. The context list should never get the long? I think we need to take a different approach.

Reply:
The context list are delayed messages (caused by frozen subtree). how large it is depends on how many messages were sent by clients while subtree was frozen.